File: doc/quote-azure-k8-backend/Services/JwtService.cs
using Microsoft.IdentityModel.Tokens;
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;
using System.Text;
using quote_azure_k8_backend.Models;
using Microsoft.Extensions.Configuration;
using quote_azure_k8_backend.Data;

namespace quote_azure_k8_backend.Services
{
    public class JwtService : IJwtService
    {
        private readonly IConfiguration _config;
        private readonly IUserRoleRepository _userRoleRepository;
        private readonly string _key;
        private readonly string _issuer;
        private readonly string _audience;

        public JwtService(IConfiguration config, IUserRoleRepository userRoleRepository)
        {
            _config = config;
            _userRoleRepository = userRoleRepository;
            _key = _config["JwtSettings:SecretKey"] ?? throw new ArgumentNullException("JwtSettings:SecretKey not configured");
            _issuer = _config["JwtSettings:Issuer"] ?? "QuoteAzureBackend";
            _audience = _config["JwtSettings:Audience"] ?? "QuoteAzureUsers";
        }

        public string GenerateToken(User user)
        {
            var tokenHandler = new JwtSecurityTokenHandler();
            var key = Encoding.ASCII.GetBytes(_key);
            
            // Get user roles from the database
            var userRole = _userRoleRepository.GetUserRoleAsync(user.Username).GetAwaiter().GetResult();
            var roleClaims = new List<Claim>();
            
            if (userRole != null)
            {
                roleClaims.Add(new Claim(ClaimTypes.Role, userRole.Role));
            }
            
            var claims = new List<Claim>
            {
                new Claim(ClaimTypes.NameIdentifier, user.Id),
                new Claim(ClaimTypes.Email, user.Email),
                new Claim(ClaimTypes.Name, user.Username),
                new Claim("jti", Guid.NewGuid().ToString()),
                new Claim("iat", DateTimeOffset.UtcNow.ToUnixTimeSeconds().ToString(), ClaimValueTypes.Integer64)
            };
            
            // Add role claims
            claims.AddRange(roleClaims);

            var tokenDescriptor = new SecurityTokenDescriptor
            {
                Subject = new ClaimsIdentity(claims),
                Expires = DateTime.UtcNow.AddMinutes(double.Parse(_config["JwtSettings:ExpiryMinutes"] ?? "60")),
                Issuer = _issuer,
                Audience = _audience,
                SigningCredentials = new SigningCredentials(new SymmetricSecurityKey(key), SecurityAlgorithms.HmacSha256Signature)
            };

            var token = tokenHandler.CreateToken(tokenDescriptor);
            return tokenHandler.WriteToken(token);
        }

        public string GenerateRefreshToken(User user)
        {
            return Guid.NewGuid().ToString();
        }

        public ClaimsPrincipal? ValidateToken(string token)
        {
            try
            {
                var tokenHandler = new JwtSecurityTokenHandler();
                var key = Encoding.ASCII.GetBytes(_key);
                
                var validationParameters = new TokenValidationParameters
                {
                    ValidateIssuerSigningKey = true,
                    IssuerSigningKey = new SymmetricSecurityKey(key),
                    ValidateIssuer = true,
                    ValidIssuer = _issuer,
                    ValidateAudience = true,
                    ValidAudience = _audience,
                    ValidateLifetime = true,
                    ClockSkew = TimeSpan.Zero
                };

                var principal = tokenHandler.ValidateToken(token, validationParameters, out SecurityToken validatedToken);
                return principal;
            }
            catch
            {
                return null;
            }
        }

        public string? GetUserIdFromToken(string token)
        {
            var principal = ValidateToken(token);
            return principal?.FindFirst(ClaimTypes.NameIdentifier)?.Value;
        }
    }
}

File: quote-api/src/main/java/com/quote/k8/service/JwtService.java
package com.quote.k8.service;

import com.quote.k8.model.User;
import com.quote.k8.model.UserRole;
import com.quote.k8.repository.UserRoleRepository;
import io.smallrye.jwt.build.Jwt;
import org.eclipse.microprofile.config.inject.ConfigProperty;

import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import java.util.Optional;

@ApplicationScoped
public class JwtService {

    @Inject
    UserRoleRepository userRoleRepository;

    @ConfigProperty(name = "mp.jwt.verify.issuer")
    String issuer;

    @ConfigProperty(name = "mp.jwt.verify.audience")
    String audience;

    public String generateToken(User user) {
        // Get user roles
        Optional<UserRole> userRole = userRoleRepository.findByUsername(user.username);
        
        var jwtBuilder = Jwt.claims()
                .subject(user.id.toString())
                .upn(user.email)
                .claim("username", user.username)
                .claim("email", user.email)
                .issuer(issuer)
                .audience(audience);
        
        // Add role if present
        userRole.ifPresent(role -> jwtBuilder.groups(role.role));
        
        return jwtBuilder.sign();
    }
}
